### DISCUSSION TITLE
How Can I get the accessToken from JavaSdk?

### DISCUSSION THREAD

**Discussion by zhang214214** [ory/hydra#2581] [Q&A] [answered]
**How Can I get the accessToken from JavaSdk?**
#################################################
ApiClient defaultClient = Configuration.getDefaultApiClient();
    defaultClient.setBasePath("http://localhost");
    
    // Configure HTTP basic authorization: basic
    HttpBasicAuth basic = (HttpBasicAuth) defaultClient.getAuthentication("basic");
    basic.setUsername("YOUR USERNAME");
    basic.setPassword("YOUR PASSWORD");

    // Configure OAuth2 access token for authorization: oauth2
    OAuth oauth2 = (OAuth) defaultClient.getAuthentication("oauth2");
    oauth2.setAccessToken("YOUR ACCESS TOKEN");

    PublicApi apiInstance = new PublicApi(defaultClient);
    String grantType = "grantType_example"; // String | 
    String code = "code_example"; // String | 
    String refreshToken = "refreshToken_example"; // String | 
    String redirectUri = "redirectUri_example"; // String | 
    String clientId = "clientId_example"; // String | 
    try {
      Oauth2TokenResponse result = apiInstance.oauth2Token(grantType, code, refreshToken, redirectUri, clientId);
      System.out.println(result);
    } catch (ApiException e) {
      System.err.println("Exception when calling PublicApi#oauth2Token");
      System.err.println("Status code: " + e.getCode());
      System.err.println("Reason: " + e.getResponseBody());
      System.err.println("Response headers: " + e.getResponseHeaders());
      e.printStackTrace();
    }
################################################
**As you can see if you going to get the token you need the token! what the hel?**

**Comment by aeneasr** [comment, marked as answer]
It is much better to use an official OAuth2 library for these interaction, see: https://www.ory.sh/hydra/docs/sdk

<img width="1005" alt="Bildschirmfoto 2021-06-21 um 16 51 27" src="https://user-images.githubusercontent.com/3372410/122782409-fddb0b00-d2b0-11eb-955a-b28b8440a003.png">

**Comment by zhang214214** [comment]
Do you have a Demo for Java to make a OAuth2.0 & OpenId Conect? I can not make it in javasdk.

**Comment by zhang214214** [comment]
![image](https://user-images.githubusercontent.com/13167133/122857404-02540200-d34b-11eb-9506-733315150dc9.png)
**I have created a client but I can not to get the accessToken**
![image](https://user-images.githubusercontent.com/13167133/122857691-768ea580-d34b-11eb-9f6f-5a0aacbeaacc.png)

**Reply by vinckr** [reply]
Hello @zhang214214 ,
sorry for the late reply! 
Could you please share a reproducible repository with shell commands (e.g. hydra clients create, curl ...) that reproduces this issue?
I will mark this as answered for now, but feel free to come back to the discussion any time.

**Reply by zhang214214** [reply]
Okey! man, I foud it because I choose the method:  "token_endpoint_auth_method":"client_secret_post" when I create the client. step like:
step 1:
**post to create  client**  
`{
    "client_id":"sharing-tommy",
    "client_name":"sharing-tommy",
    "client_secret":"55e7315d-0c85-4eef-a5b3-cef6bf217351",
    "client_secret_expires_at":0,
    "redirect_uris": [
        "https://oa.tommy-test.cn"
  ],
  "post_logout_redirect_uris": [
      "https://oa.tommy-test.cn/logout"
  ],
    "created_at":"2020-01-06T15:09:15.946Z",
    "frontchannel_logout_session_required":true,
    "frontchannel_logout_uri": "https://oa.tommy-test.cn/callback",
    "scope":"openid offline offline_access",
    "token_endpoint_auth_method":"client_secret_post",
    "updated_at":"2020-01-07T15:09:15.946Z",
    "userinfo_signed_response_alg":"none",
    "grant_types": [
    "authorization_code","refresh_token","implicit","client_credentials"
  ],
   "response_types": [
    "code","id_token","token"
  ]
}`

step 2 :
**get the token**
`    
    
    public HydraToken getOAuth(String code) {

        OkHttpClient client = new OkHttpClient();

        RequestBody formBody = new FormBody.Builder()
                .add("grant_type", "authorization_code")
                .add("code", code)
                .add("client_id", "sharing-tommy")
                .add("client_secret", "55e7315d-0c85-4eef-a5b3-cef6bf217351")
                .add("redirect_uri", "https://oa.tommy-test.cn")
                .build();

        Request request = new Request.Builder()
                .url(hydraServer + "/oauth2/token")
                .post(formBody)
                .build();

        try {
            Response response = client.newCall(request).execute();
            JsonNode jsonNode = mapper.readTree(response.body().string());
            if (StringUtils.isNotBlank(jsonNode.path("access_token").asText())) {
                return JSON.parseObject(jsonNode.toString(), HydraToken.class);
            }

        } catch (IOException e) {
            e.printStackTrace();
        }

        return null;
    }`

**Reply by zhang214214** [reply]
And now I can get token by Java

**Comment by allenhaozi** [comment]
```go
                 code := r.URL.Query().Get("code")
		response["code"] = code

		params := &public.Oauth2TokenParams{}

		clientID := "client-id-01"
		redirectURI := "http://localhost:9999/callback"

		params.SetClientID(&clientID)
		params.SetCode(&code)
		params.WithContext(ctx)
		params.SetGrantType("authorization_code")
		params.SetRedirectURI(&redirectURI)

		res, err := s.HydraPublicClient.Public.Oauth2Token(params,nil)
		
```
got error
```
[POST /oauth2/token][401] oauth2TokenUnauthorized  &{Error:invalid_client ErrorDebug: ErrorDescription:Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method). StatusCode:0}"
```

What went wrong?

**Reply by allenhaozi** [reply]
Ignore the problem. It's been solved

step 1:
          change token_endpoint_auth_method from client_secret_post to client_secret_basic

step 2:

```go
info := openapiclient.BasicAuth("${client-id}", ""${client-secret})
res, err := s.HydraPublicClient.Public.Oauth2Token(params,info)
```

**Reply by vinckr** [reply]
Thanks a ton for providing your solution @allenhaozi !
